### PR TITLE
Extract UI into separate repo

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

We have decided to extract the new `ui/` directory out into [its own repo](https://github.com/guardian/frontend-future), in order to decouple it from `frontend`. We have published it to NPM using the Guardian org under [`@guardian/guui`](https://www.npmjs.com/package/@guardian/guui)

### TODO
 - Provide a seamless process allowing developers to point their local `frontend` instance at their local version of `@guardian/guui`
- When a new version of `@guardian/guui` is published, it would be great if a corresponding PR was created against `frontend`, updating to the latest version, like a mini-[Greenkeeper](https://greenkeeper.io/)

Any other suggestions or things we haven't considered would be welcome!

@guardian/dotcom-platform 

## What is the value of this and can you measure success?

👍  Quicker and less complicated validation steps in `frontend`
👍  Less coupling between the new `ui/` and the current rendering and asset management layer

👎  Changes to `@guardian/guui` will require the package to be published to NPM _and_ `frontend` to be redeployed before the changes are picked up.

## Does this affect other platforms - Amp, Apps, etc?

No

## Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

No

## Screenshots

No

## Tested in CODE?

No

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@guardian/guui": "^0.1.0-alpha.7",
```
Comment (gustavpursche) [inline package.json]: Should we simply point it to `@latest` for now?

Comment (sndrs) [inline package.json]: I think you'd have to have to `yarn upgrade @guardian/guui` in both cases?

Comment (sndrs) [inline package.json]: but `@latest` may be a truer reflection of how they relate?

**Comment by TBonnin** [comment]
Is this temporary?

Having 2 repositories adds a lot of complexity imho:

- In term of development experience, devs needs to know there are multiple repos, clone and keep them in sync and learn how to point the dev Play app to the ui dev environment.
- Every change touching the UI is now requiring 2 pulls requests and 2 reviews.
- It also mean there are more moving parts, adding additional steps to the CI process (probably 2 projects/config).
- Deployment is becoming a 2 phases process where you need to publish the new UI first and then deploy the app to use the new UI. We would also need to take care of potential mistakes where the play app is deployed first (or when publishing the ui fails).
- I don't think Riff-Raff support npm package publication, which mean we either need to add it or release frontend UI in a way nobody else do it in the department. It would also mean we can not use Riff-Raff to orchestrate the dependencies between our different deployments.
- Grep'ing for code is becoming trickier, if you don't know already what you are looking for you have to search in 2 repos.

Does the benefit of having 2 repos is worth the amount of friction this is adding?

**Comment by SiAdcock** [comment]
@TBonnin thanks for your questions!

I agree with your points about the added complexity. We can manage some of your concerns to a certain extent, although whether the pros outweigh the cons is difficult to quantify. This extraction can be seen as a trial to see if having a separate UI repo creates too much friction.

> devs needs to know there are multiple repos, clone and keep them in sync and learn how to point the dev Play app to the ui dev environment.

> Grep'ing for code is becoming trickier, if you don't know already what you are looking for you have to search in 2 repos.

The best way to address these concerns is through publicity: documentation, tech time talks and client side meetups. It is not too different to the way other consumable libraries work, such Ophan's tracker JS and the forthcoming universal Atoms that journalism is working on. Eventually, all pages will be built using components composed by `@guardian/guui`, which will make it easier to find the UI code you're looking for. Client side support during the transitional phase will be important.

> Every change touching the UI is now requiring 2 pulls requests and 2 reviews.

Agreed, although the PR against frontend could be created automatically, and the code review would be fairly superficial. 

> Deployment is becoming a 2 phases process where you need to publish the new UI first and then deploy the app to use the new UI. We would also need to take care of potential mistakes where the play app is deployed first (or when publishing the ui fails).

If the PR against `frontend` is conditional on `@guardian/guui` being published, we can alleviate concerns about what happens if the Ui publish fails

> I don't think Riff-Raff support npm package publication, which mean we either need to add it or release frontend UI in a way nobody else do it in the department. It would also mean we can not use Riff-Raff to orchestrate the dependencies between our different deployments.

I have discussed using Riff Raff to publish to NPM with Simon Hildrew in the past. He was against the idea as it adds this third party state (the NPM repository) that Riff Raff need to know about, making deployment brittle and non-idempotent. I tend to agree with him. Riff Raff is designed to deploy artefacts, not to build and publish libraries.

@sndrs and I have discussed using Travis CI to automate the publication of `@guardian/guui`. It is fairly trivial to set this up. We already publish `mobile-apps-article-templates` to NPM in a similar way, using Circle CI. Alternatively we can look into doing this in TeamCity if we want to be consistent with `frontend` CI, although I'm wary that TeamCity configuration is overly complex for such a small and vanilla task.

Happy to discuss any of this further / face to face 😄

**Comment by sndrs** [comment]
just to add to @SiAdcock's points:

it's not intended to be explicitly temporary, but it if makes life harder then it will be. it's also pretty simple to add it back if that's needed.

currently, we are finding that colocating the code is making frontend devex bloated and hard to manage, and we expect that situation to get worse as more code is written. 

our assumption is that working on the guui repo is unlikely, in practice, to often require edits to the rest of frontend (that seems to be how we have worked on this stuff so far), and that isolating the UI code will make working on both repos lighter and quicker. 

again, if this assumption is wrong, we will revert this change.

in terms of devex, we anticipate:

- raise PR on UI repo
- build checked by travis (tests, perf etc)
- +1 and merge to master
- travis versions and publishes updated package to npm
- travis raises PR in frontend bumping UI version in `package.json`/`yarn.lock`, cc'ing the user that merged the PR on the UI repo
- user verifies on CODE and merges frontend

we think this will mean:

- frontend and UI builds will both be lighter, less complex and quicker to green light, since neither will be testing (probably) unchanged code from the other
- the UI CI will be easier to maintain (a yml file vs AMIs/agents) while frontend will be unaffected

the cost of this will be a 2nd PR, which normally will be trivial to review (we might consider it already reviewed in the initial UI PR if it's just a version bump). 

as with all UI stuff, this is a hypothesis that life will be better this way. if it's not, we'll make another and try that

**Comment by TBonnin** [comment]
> currently, we are finding that colocating the code is making frontend devex bloated and hard to manage, and we expect that situation to get worse as more code is written.

I hear you. This is definitely a problem worth solving.
However the 2 repos solution is only moving what is a core problem (ie: how to build/lint,etc... assets) that should be transparent to 80% of the department into a global problem where each devs needs to keep in mind there are now more moving parts.

> our assumption is that working on the guui repo is unlikely, in practice, to often require edits to the rest of frontend (that seems to be how we have worked on this stuff so far)

I disagree. There is plenty of PRs adding stuff to the UI that requires to pass the data to display to the template. All of those PRs that touches the interface between Play and the UI would requires 2 pull requests.

> raise PR on UI repo
> build checked by travis (tests, perf etc)
> +1 and merge to master
> travis versions and publishes updated package to npm
> travis raises PR in frontend bumping UI version in package.json/yarn.lock, cc'ing the user that merged the PR on the UI repo
> user verifies on CODE and merges frontend

That's a very simplified dev flow. A few initial steps are missing:
1. Checkout 2 repos
2. Point Play app to local UI dev
3. Implement your change working in 2 directories
3. Potentially making 2 set of commits and PRs on both repos
...etc (if things go wrong they can go wrong in 2 different places)

Bumping automatically the UI version in frontend package.json only works if we are never touching the Play/UI interface. When passing new data to the UI components for example we want to manually specify which version of the UI is required. Doing this automatically would lead to issues where the UI expects data the app is not passing yet, because both deploy would not be coordinated anymore.

Let's discuss face to face (maybe a meeting) to find the best possible solution to this problem.

**Comment by TBonnin** [comment]
As mentioned during our call, I would prefer to bite the bullets now and fix the reasons that pushes us to move away from the frontend repo: 
- Quicker feedback loop 
- More flexible TC (ie: nvm)
- More efficient and faster git hooks

However that's up to you, as long as it is transparent for the rest of the developers.

**Comment by sndrs** [comment]
i'm persuaded by your arguments @TBonnin – if things are such that it's easier not to add code (or feels it) to frontend, we should sort out frontend. that's a fair point

**Comment by mario-galic** [comment]
image 
Such beauty :heart_eyes:... Can we employ a full time developer dedicated to only removing code?

**Comment by sndrs** [comment]
shall we close this?

**Pull request closed without merging**